### ISSUE TITLE
How do I read several subsequent "messages" from InflaterSource

### ISSUE THREAD

**Issue by AntonRutkevich** [square/okio#570]
Hi! 

### TLDR;
How do I read several subsequent "messages" from InflaterSource? 
How do I make the test below pass without throwing the EOFException?

Okio version: 1.17.2

I'm currently implementing support for websocket "permessage-deflate" compression in okhttp, with and without context takeover. Putting aside the [octet removal/adding](https://tools.ietf.org/html/rfc7692#section-7.2.1), the general idea as I see it is to create a single `DeflaterSink` and a single `InftalerSource` per `WebSocket`. 
Then, if "server_no_context_takeover" is **not** in the "agreed parameters", DeflaterSink's underlying `java.util.zip.Deflater` keeps the context during the entire connection, allowing much better compression level.

So, a single deflate-inflate cycle looks like:

```
sender: 
	deflate message
	flush
reciever:
	inflate deflated message
```

The issue I have is with inflating the previously deflated message. The `InflaterSource` throws `java.io.EOFException: source exhausted prematurely` exception from `Buffer().writeAll(inflaterSource)` call. If I just catch and ignore that exception, everything works fine, but I doubt it's the right way to go.

Here is the full source code with the test.

```kotlin
package okhttp3.internal.ws

import okio.Buffer
import okio.DeflaterSink
import java.io.Closeable
import java.io.IOException
import java.util.zip.Deflater

class MessageDeflater : Closeable {
  private val sink = Buffer()
  private val deflater = Deflater(Deflater.BEST_COMPRESSION, true /* omit zlib header */)
  private val deflaterSink = DeflaterSink(sink, deflater)

  @Throws(IOException::class)
  fun deflate(source: Buffer, contextTakeover: Boolean): Buffer {
    if (!contextTakeover) {
      deflater.reset()
    }

    source.readAll(deflaterSink)
    deflaterSink.flush()

    return sink.apply {
      println("Deflated: ${snapshot().hex()}")
    }
  }

  @Throws(IOException::class)
  override fun close() {
    deflater.end()
  }
}
```

```kotlin
package okhttp3.internal.ws

import okio.Buffer
import okio.InflaterSource
import java.io.Closeable
import java.io.EOFException
import java.io.IOException
import java.util.zip.Inflater

class MessageInflater : Closeable {
  private val source = Buffer()
  private val inflater = Inflater(true /* omit zlib header */)
  private val inflaterSource = InflaterSource(source, inflater)

  @Throws(IOException::class)
  fun inflate(deflated: Buffer): Buffer {
    source.writeAll(deflated)

    return Buffer().apply {
      try {
        writeAll(inflaterSource)
      } catch (e: EOFException) {
        e.printStackTrace()
      }
    }
  }

  @Throws(IOException::class)
  override fun close() {
    inflater.end()
  }
}
```

```java
package okhttp3.internal.ws;

import org.junit.Test;

import java.io.IOException;

import okio.Buffer;
import okio.ByteString;

import static org.assertj.core.api.Assertions.assertThat;

public final class MessageDeflaterTest {

  private MessageDeflater deflater = new MessageDeflater();
  private MessageInflater inflater = new MessageInflater();

  @Test
  public void twoSubsequentTextsDeflatedInflated() throws IOException {
    String src = "Hello deflate!";
    Buffer deflated = deflater.deflate(new Buffer().writeUtf8(src), true);
    ByteString inflated = inflater.inflate(deflated).readByteString();

    assertThat(inflated.utf8()).isEqualTo(src);

    String src2 = "Hello deflate!";

    Buffer deflated2 = deflater.deflate(new Buffer().writeUtf8(src2), true);
    ByteString inflated2 = inflater.inflate(deflated2).readByteString();
    assertThat(inflated2.utf8()).isEqualTo(src2);
  }
}
```

The test above PASSES with the output:
```
Deflated: f248cdc9c95748494dcb492c495504000000ffff
java.io.EOFException: source exhausted prematurely
	at okio.InflaterSource.read(InflaterSource.java:83)
	at okio.Buffer.writeAll(Buffer.java:1143)
	at okhttp3.internal.ws.MessageInflater.inflate(MessageInflater.kt:21)
	at okhttp3.internal.ws.MessageDeflaterTest.twoSubsequentTextsDeflatedInflated(MessageDeflaterTest.java:21)
	...

Deflated: f240e101000000ffff
java.io.EOFException: source exhausted prematurely
	at okio.InflaterSource.read(InflaterSource.java:83)
	at okio.Buffer.writeAll(Buffer.java:1143)
	at okhttp3.internal.ws.MessageInflater.inflate(MessageInflater.kt:21)
	at okhttp3.internal.ws.MessageDeflaterTest.twoSubsequentTextsDeflatedInflated(MessageDeflaterTest.java:28)
	...
```

If `flush()` in `MessageDeflater` is replaced with `close()`, the first "message" is deflated with slightly different bytes, inflated without any exception, but subsequent messages can't be deflated obviously, as the `Deflater` has been closed:

```
Deflated: f348cdc9c95748494dcb492c49550400

java.lang.NullPointerException: Deflater has been closed

	at java.util.zip.Deflater.ensureOpen(Deflater.java:559)
	at java.util.zip.Deflater.reset(Deflater.java:522)
	at okhttp3.internal.ws.MessageDeflater.deflate(MessageDeflater.kt:17)
	at okhttp3.internal.ws.MessageDeflaterTest.twoSubsequentTextsDeflatedInflated(MessageDeflaterTest.java:40)
	...
```

Any idea how to get this working?

**Comment by swankjesse** [comment]
You don't want readAll() / writeAll(), these methods assume the stream terminates (ie. close, not flush) but yours do not.

We had to do something similar with SPDY/3 headers. Our strategy was to read from an InflaterSource, stopping when we knew there was no more bytes to read. Do you get any side channel information on how many uncompressed bytes you'll get? If you do, just read that many bytes. Otherwise you need to somehow get the inflater to tell you if it has bytes before you attempt to read them.

https://github.com/square/okhttp/blob/okhttp_27/okhttp/src/main/java/com/squareup/okhttp/internal/framed/NameValueBlockReader.java

https://github.com/square/okhttp/blob/okhttp_27/okhttp/src/main/java/com/squareup/okhttp/internal/framed/Spdy3.java#L408

**Comment by AntonRutkevich** [comment]
Thanks @swankjesse, that helped a lot!

I was able to track the total number of incoming deflated bytes and then compare them with the number of incoming bytes processed by `Inflater`. You can find the implementation in the okhttp PR: 
https://github.com/square/okhttp/pull/4871/files#diff-8fe92fe7899eb8a76a72456f82232a6c

I believe this issue can be closed and we can discuss further in the above okhttp PR.

**Comment by swankjesse** [comment]
# 💥

Awsome
